utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -264,32 +276,38 @@ class BaseDeltaIterator : public Iterator {
       } else if (!delta_iterator_->status().ok()) {
         // Expose the error status and stop.
         current_at_base_ = false;
-        return;
+        break;
       }
       equal_keys_ = false;
       if (!BaseValid()) {
         if (!base_iterator_->status().ok()) {
           // Expose the error status and stop.
           current_at_base_ = true;
-          return;
+          break;
         }
 
         // Base has finished.
         if (!DeltaValid()) {
           // Finished
-          return;
+          break;
         }
         if (delta_entry.type == kDeleteRecord ||
             delta_entry.type == kSingleDeleteRecord) {
           AdvanceDelta();
+          // If the new Delta is valid and >= iterate_upper_bound, stop
+          current_over_upper_bound_ =
+              DeltaValid() && IsOverUpperBound(delta_iterator_->Entry().key);
+          if (current_over_upper_bound_) {
+            return;
+          }
```

Comment:
Do we consider to do the check inside `AdvanceDelta()` instead?

Reply:
`delta_iterator_` modification comes from two paths: 
1) variants of Seek, rely on `UpdateCurrent` to refresh `current_over_upper_bound_`
2) Next/Prev, refresh `current_over_upper_bound_` through `AdvanceDelta` 
If we keep the current code structure, we cannot remove the check from `UpdateCurrent()`